Assets/Implementation/Pathfinding/Scripts/NPCController.cs

```diff
@@ -0,0 +1,213 @@
+using System.Collections;
+using System.Collections.Generic;
+using Unity.Collections;
+using Unity.Jobs;
+using UnityEngine;
+using Grid = Features.Building.Scripts.Grid.Grid;
+
+namespace Implementation.Pathfinding.Scripts
+{
+public class NPCController : MonoBehaviour
+{
+    [Header("Dependecies")]
+    [SerializeField] private Grid _grid;
+    [SerializeField] private bool _drawGizmos;
+
+    [Header("Movement variables")]
+    [SerializeField] float _moveSpeed = 1f;
+
+    Node[,,] _nodeGrid;
+    List<Node> _backtrackedPath = new List<Node>();
+    [SerializeField] Vector3Int _endNode;
+    int _gridWidth;
+    int _gridHeight;
+    [SerializeField] bool _pathCompleted = true;
+    [SerializeField] Node[] open;
+    [SerializeField] Node[] closed;
+
+    public bool PathCompleted => _pathCompleted;
+
+    private void Start()
+    {
+        _gridWidth = _grid.GridSize.x;
+        _gridHeight = _grid.GridSize.y;
+    }
+
+    public Vector3Int GetGridPosition()
+    {
+        var pos = transform.position;
+
+        //clamp to grid positions
+        var clampedValue = new Vector2(RoundToMultiple(pos.x, _grid.CellSize),
+            RoundToMultiple(pos.y, _grid.CellSize));
+
+        return new Vector3Int((int)clampedValue.x, (int)clampedValue.y, 0);
+    }
+
+    void Update()
+    {
+        var pos = Camera.main.ScreenToWorldPoint(Input.mousePosition);
+
+        //clamp to grid positions
+        var clampedValue = new Vector2(RoundToMultiple(pos.x, _grid.CellSize),
+            RoundToMultiple(pos.y, _grid.CellSize));
+
+        if (Input.GetKeyDown(KeyCode.P) & _pathCompleted)
```

Comment:
Hardcoded input control. Move this to a demo file and control it with public functions